### DISCUSSION TITLE
Interoperability with FastRTPS

### DISCUSSION THREAD

**Discussion by nbrow100** [OpenDDS/OpenDDS#4627] [Q&A]
Any message that gets fragmented between fastrtps (version 2.9.1, rtps 2.3) and OpenDDS (version 3.23.1, rtps 2.5) doesn't get handled by the subscriber. 

Below image is a large message getting split into 3 message fragments from the fastrtps side (rtps data_frag). The OpenDDS side receives the message (see attached log), but drops the sample complaining about the CDR/XCDR. With the exact same QoS settings, this won't happen with unfragmented messages.

_(1919|1986) navwar_dds::idl::SortedDataDataReaderImpl::dds_demarshal object construction failure, dropping sample.
(1919|1980) ERROR EncapsulationHeader::to_encoding: Unexpected Extensibility Encoding: CDR/XCDR1 Big Endian Plain_

Digging through the message (attached), I find that the writerSeqNumber in the fragmented message (data_frag) is 8 bytes long. OMG seems to specify a 4 byte writerSeqNumber in the data_frag message. 

![image](https://github.com/OpenDDS/OpenDDS/assets/20782471/2c17b31b-3aee-4345-8fc6-434b80339933)
[opendds_message_log.txt](https://github.com/OpenDDS/OpenDDS/files/12479854/opendds_message_log.txt)
[exported_fragmented_packets.txt](https://github.com/OpenDDS/OpenDDS/files/12479895/exported_fragmented_packets.txt)

[common]
DCPSGlobalTransportConfig=rtps
DCPSDefaultDiscovery=DEFAULT_RTPS
DCPSBit=1
DCPSDebugLevel=5
DCPSLogLevel=debug
DCPSTypeObjectEncoding=ReadOldFormat

[config/rtps]
transports=the_rtps_transport
swap_bytes=1

[transport/the_rtps_transport]
transport_type=rtps_udp
max_message_size=655510
rcv_buffer_size=6553600

**Comment by mitza-oci** [comment]
The Sequence Number is always 8 bytes.  It identifies the data sample that existed before fragmentation.  There is an additional 4 byte fragment number.  See "9.4.5.5 DataFrag Submessage" in RTPS spec.

**Comment by nbrow100** [comment]
![image](https://github.com/OpenDDS/OpenDDS/assets/20782471/0a2d521c-21fd-4812-a9fd-c8d4b5e7917b)
Thanks for the quick reply, I was referencing this guy from the rtps spec. Maybe I'm looking at it wrong.

**Comment by mitza-oci** [comment]
Yes, that extra `+` in the margin counts for the other 4 bytes.  More formally, it's in "9.4.2.5 SequenceNumber" which includes both IDL and ASCII art.

**Comment by nbrow100** [comment]
Well that makes sense, appreciate you clearing that up for me. I'm a dummy.  I didn't notice the + in the margin. Other messages from that spec seemed accommodate the extra bytes with more white space!


In that case, I have no clue why it's not working other than a xcdr/cdr qos settings we haven't figured out.

**Comment by mitza-oci** [comment]
Could you attach a packet capture (pcap/pcapng file) of both fragmented and unfragmented data generated by Fast RTPS?

**Comment by nbrow100** [comment]
[test.zip](https://github.com/OpenDDS/OpenDDS/files/12487511/test.zip)
This includes unfragmented and fragmented messages. The unfragmented message is 60kb, the fragmented message is about 148kb or 2 65kb message and a 17kb messages.

**Comment by jrw972** [comment]
The encapsulation kind in the first fragment (packet 58) is CDR_LE (0x0001).
The encapsulation kind in the second fragment (packet 59) is CDR_BE (0x0000).
The third fragment is also CDR_BE.

I'm not an expert on the spec, but it seems like all fragments should have the same encapsulation kind.  This explains the log

    (1919|1980) ERROR EncapsulationHeader::to_encoding: Unexpected Extensibility Encoding: CDR/XCDR1 Big Endian Plain

**Comment by mitza-oci** [comment]
I don't think the other (after first) fragments have an encapsulation kind.

**Comment by jrw972** [comment]
You're right.  That's wireshark peeking inside the fragment data.

**Comment by nbrow100** [comment]
I only see encapsulation in the first fragment. CDR_LE. All of our data reports as CDR_LE I thought. 

To be honest I'm confused why it's CDR_LE but that's a knowledge gap I have between code and what's actually going on, so I apologize. Our QoS settings are set as XCDR_DATA_REP which has a value of '0' which  I thought should be CDR_BE to start with. But every packet whether its fragmented or not shows as CDR_LE. So I may not be setting that right in code.


![image](https://github.com/OpenDDS/OpenDDS/assets/20782471/bba56a97-987c-4719-9751-c54a57f9f26a)
![image](https://github.com/OpenDDS/OpenDDS/assets/20782471/949e5b1f-ab27-4bce-bb0f-99b42e9c9d9e)

![image](https://github.com/OpenDDS/OpenDDS/assets/20782471/9e793105-05a4-4bd5-a430-aaa36414dab4)

**Comment by nbrow100** [comment]
As an aside, my current version of wireshark (Version 4.0.8 (v4.0.8-0-g81696bb74857)) doesn't always peak at rtps correctly. A lot of the messages that do work on my local interface don't get captured if they are too large. For instance if I send a sufficiently large messsage that gets fragmented OpenDDS->OpenDDS(rtps) or FastDDS->FastDDS(rtps), it doesn't get picked up by my wireshark but the data does flow between producer and consumer.

 So there is probably a ton of discrepancies at how wireshark reports CDR or looks inside messages. Just what I've observed. I don't think it contributes to this conversation much. I appreciate all the help and responses thus far.

**Comment by jrw972** [comment]
Based on the log, it appears you are using OpenDDS 3.23 instead of 3.23.1.
There are at least four areas to explore
1. The data was corrupt in the fragments.
2. Reassembling the fragments corrupted the data.
3. The IDL is different for the writer and reader.
4. There is an encoding mismatch where the DataWriter is writing using one encoding while the DataReader expects a different one.

For (1), you could serialize (write) the sample using OpenDDS or some other DDS to ensure it was serialized correctly.
Similarly, for (2), you could use an OpenDDS writer with an OpenDDS reader and compare the fragmentation.

For (3) and (4), you would need to check the IDL, [Data Representation](https://opendds.readthedocs.io/en/latest-release/devguide/xtypes.html#data-representation), and extensibility.  For example, the writer could be writing CDR (XCDR1) while the reader expects XCDR2.  Or perhaps the IDL for the writer indicates a mutable struct while the IDL for the reader indicates a final struct.

**Comment by mitza-oci** [comment]
> I only see encapsulation in the first fragment. CDR_LE. All of our data reports as CDR_LE I thought.
> 
> To be honest I'm confused why it's CDR_LE but that's a knowledge gap I have between code and what's actually going on, so I apologize. Our QoS settings are set as XCDR_DATA_REP which has a value of '0' which I thought should be CDR_BE to start with. But every packet whether its fragmented or not shows as CDR_LE. So I may not be setting that right in code.

These are different but related.  The QoS can be used to select the CDR version XCDR1 or XCDR2 (from XTypes) or some other implementation-specific encoding types.  The constants like CDR_BE also specify the endianness which is not configurable from QoS.

**Comment by mitza-oci** [comment]
> As an aside, my current version of wireshark (Version 4.0.8 (v4.0.8-0-g81696bb74857)) doesn't always peak at rtps correctly. A lot of the messages that do work on my local interface don't get captured if they are too large. For instance if I send a sufficiently large messsage that gets fragmented OpenDDS->OpenDDS(rtps) or FastDDS->FastDDS(rtps), it doesn't get picked up by my wireshark but the data does flow between producer and consumer.
> 
> So there is probably a ton of discrepancies at how wireshark reports CDR or looks inside messages. Just what I've observed. I don't think it contributes to this conversation much. I appreciate all the help and responses thus far.

Yes, Wireshark's dissector is not always 100% working and matching the spec.  We have contributed some changes to it in the past but I'm sure there's more to do.

The raw bytes of the UDP payload should always be available in Wireshark since that doesn't depend on the dissector.  Sometimes we resort to just manually parsing through it.

**Comment by lozanoeg** [comment]
> The QoS can be used to select the CDR version XCDR1 or XCDR2 (from XTypes) or some other implementation-specific encoding types. The constants like CDR_BE also specify the endianness which is not configurable from QoS.

I ended up working on this as well and was able to convert everything over to BE  as well as confirming XCDR1 and turned on the Transport Debugging
I can see that the OpenDDS DataReader receives the fragments from The FastDDS DataWriter however it does not reassemble them, the outputs i am seeing looking like below. the unexpected extensibility seems to happen only when i disconnect the data writer now and not on each fragment
```
(107397|107426) INFO: RtpsUdpReceiveStrategy::handle_input: reallocating primary receive buffer based on reference count
(107397|107426) INFO: RtpsUdpReceiveStrategy::handle_input: reallocating primary receive buffer based on reference count
(107397|107426) INFO: RtpsUdpReceiveStrategy::handle_input: reallocating primary receive buffer based on reference count
(107397|107426) INFO: RtpsUdpReceiveStrategy::handle_input: reallocating primary receive buffer based on reference count
(107397|107426) INFO: RtpsUdpReceiveStrategy::handle_input: reallocating primary receive buffer based on reference count
(107397|107426) INFO: RtpsUdpReceiveStrategy::handle_input: reallocating primary receive buffer based on reference count
(107397|107426) INFO: RtpsUdpReceiveStrategy::handle_input: reallocating primary receive buffer based on reference count
(107397|107417) RtpsSampleHeader::into_received_data_sample() - used KeyHash data as the key-only payload
(107397|107417) ERROR EncapsulationHeader::to_encoding: Unexpected Extensibility Encoding: CDR/XCDR1 Big Endian Plain
(107397|107417) RtpsSampleHeader::into_received_data_sample() - used KeyHash data as the key-only payload
(107397|107417) ERROR EncapsulationHeader::to_encoding: Unexpected Extensibility Encoding: CDR/XCDR1 Big Endian Plain
(107397|107417) RtpsSampleHeader::into_received_data_sample() - used KeyHash data as the key-only payload
(107397|107416) DataReaderImpl::remove_associations: bit 0 local 0103001c.42774fe4.a385f29c.00000004(cb14de69) remote 010fb336.d2a34b6e.00000000.00000103(4ff21486) num remotes 1
(107397|107416) DataReaderImpl::remove_associations_i: bit 0 local 0103001c.42774fe4.a385f29c.00000004(cb14de69) remote 010fb336.d2a34b6e.00000000.00000103(4ff21486) num remotes 1
```

**Comment by lozanoeg** [comment]
would this mean when we generate our idl we would need to use the final extensibility or would it be better to use the --old-typeobject-encoding flag when generating

**Comment by mitza-oci** [comment]
> would this mean when we generate our idl we would need to use the final extensibility or would it be better to use the --old-typeobject-encoding flag when generating

I don't think we can answer that until we know what's causing the problem.  Neither of those sound to me like something that would only impact fragmented data samples.

**Comment by jrw972** [comment]
If the unexpected extensibility only occurs when the writer is disconnect, then it is probably related to RTPS discovery like the unregister/dispose message for the publication.

I would start with trying final extensibility.

**Comment by lozanoeg** [comment]
I tried the final extensibility as well and no luck. it's almost like it's not using XCDR1 even when i add the idl flag to force it to that encoding as well as putting it in the QOS but again everything works fine until the fragmentation of the packet larger than 65000 bytes

**Comment by lozanoeg** [comment]
Also we tried to get the pcap of the OpenDDS -> OpenDDS but it does not appear to capture any of the fragmentation packets like it does with FastDDS

**Comment by jrw972** [comment]
Did the OpenDDS -> OpenDDS configuration transfer the data correctly?  If so, there must either be a DATA or DATA_FRAG.  If it didn't fragment, then maybe you need to set a `max_message_size` to force fragmentation.

Can you share the IDL?

**Comment by mitza-oci** [comment]
> Also we tried to get the pcap of the OpenDDS -> OpenDDS but it does not appear to capture any of the fragmentation packets like it does with FastDDS

Which adapter(s) are you capturing data from in wireshark?

**Comment by lozanoeg** [comment]
> Which adapter(s) are you capturing data from in wireshark?

Capturing on loopback and the ethernet adapter

**Comment by lozanoeg** [comment]
> Did the OpenDDS -> OpenDDS configuration transfer the data correctly? If so, there must either be a DATA or DATA_FRAG. If it didn't fragment, then maybe you need to set a `max_message_size` to force fragmentation.
> 
> Can you share the IDL?

I can't share it but yes it works OpenDDS -> OpenDDS, OpenDDS -> FastDDS but not FastDDS -> OpenDDS

**Comment by mitza-oci** [comment]
I ran this test on Ubuntu 22.04 to verify that the DATA_FRAG submessages are captured:

In the background or a different session:
`sudo tcpdump -w /tmp/largesample.pcap -i any udp`

```
cd tests/DCPS/LargeSample
./run_test.pl rtps
# stop the tcpdump packet capture
tshark -Y rtps -r /tmp/largesample.pcap | grep DATA_FRAG
# (84 lines of output, yours may differ if there were retries)
```

The same test should work with the Wireshark GUI I just thought it would be easier to list the steps from the command line.

**Comment by lozanoeg** [comment]
I'll give it a shot. we are locked into 20.04 but i dont think that will change anything
